File: dotnet/src/Envdoctor/Commands/SnapshotCommand.cs
using Envdoctor.Core;
using Envdoctor.Runtime;
using Envdoctor.Utils;

namespace Envdoctor.Commands;

public sealed class SnapshotArgs
{
    public string? Root { get; set; }
    public string? Output { get; set; }
    public bool Token { get; set; }
    public bool Json { get; set; }
    public bool Globals { get; set; }
}

public static class SnapshotCommand
{
    /// `envdoctor snapshot` — capture this machine's live runtime.
    public static int Run(SnapshotArgs args)
    {
        var root = Path.GetFullPath(args.Root ?? ".");

        var snapshot = Capture.CaptureSnapshot(args.Globals);

        if (args.Output is { } output)
        {
            var dest = Path.Combine(root, output);
            File.WriteAllText(dest, SnapshotJsonPretty(snapshot) + "\n");
            Console.Error.WriteLine($"✓ Snapshot written to {output}");
        }

        if (args.Json)
        {
            Console.Out.WriteLine(SnapshotJsonPretty(snapshot));
            return ExitCodes.ExitOk;
        }

        if (args.Token)
        {
            Console.Out.WriteLine(Runtime.Token.EncodeToken(snapshot));
            return ExitCodes.ExitOk;
        }

        // Human summary.
        const string title = "RUNTIME SNAPSHOT";
        Console.Out.WriteLine(title);
        Console.Out.WriteLine(new string('─', title.Length * 2) + "\n");
        Console.Out.WriteLine($"  OS  {snapshot.Os.Platform}/{snapshot.Os.Arch} {snapshot.Os.Release}\n");

        Console.Out.WriteLine("  Tools");
        if (snapshot.Tools.Count == 0)
        {
            Console.Out.WriteLine("  none detected");
        }
        else
        {
            foreach (var t in snapshot.Tools)
                Console.Out.WriteLine($"  ✓ {t.Tool,-8} {t.Version}  {t.ResolvedFrom}");
        }

        Console.Out.WriteLine($"\n  PATH ({snapshot.Path.Count} entries)");
        var index = 0;
        foreach (var p in snapshot.Path.Take(12))
        {
            index++;
            Console.Out.WriteLine($"  {index,2}  {p}");
        }
        if (snapshot.Path.Count > 12)
            Console.Out.WriteLine($"  … {snapshot.Path.Count - 12} more");

        var ecosystems = snapshot.Globals.Keys.ToList();
        if (ecosystems.Count > 0)
        {
            Console.Out.WriteLine("\n  Globals");
            foreach (var eco in ecosystems)
                Console.Out.WriteLine($"  {eco}: {snapshot.Globals[eco].Count} packages");
        }
        else if (!args.Globals)
        {
            Console.Out.WriteLine("\n  Globals omitted — pass --globals to include the package inventory.");
        }

        Console.Out.WriteLine("\n  Share with:  envdoctor snapshot --token   ·   compare with:  envdoctor snapshot-diff <a> <b>");

        return ExitCodes.ExitOk;
    }

    private static string SnapshotJsonPretty(Models.RuntimeSnapshot snapshot)
    {
        // Pretty-print the same shape the token codec emits compactly.
        var compact = Runtime.Token.SnapshotToJson(snapshot);
        using var doc = System.Text.Json.JsonDocument.Parse(compact);
        return Json.Pretty(ToPlain(doc.RootElement));
    }

    private static object? ToPlain(System.Text.Json.JsonElement el) => el.ValueKind switch
    {
        System.Text.Json.JsonValueKind.Object => el.EnumerateObject()
            .Aggregate(new JsonObject(), (obj, p) =>
            {
                obj.Add(p.Name, ToPlain(p.Value));
                return obj;
            }),
        System.Text.Json.JsonValueKind.Array => el.EnumerateArray().Select(ToPlain).ToList(),
        System.Text.Json.JsonValueKind.String => el.GetString(),
        System.Text.Json.JsonValueKind.Number => el.TryGetInt64(out var l) ? l : el.GetDouble(),
        System.Text.Json.JsonValueKind.True => true,
        System.Text.Json.JsonValueKind.False => false,
        _ => null,
    };
}

File: swift/Sources/EnvdoctorKit/Commands/SnapshotCommand.swift
import Foundation

public struct SnapshotArgs {
    public var root: String?
    public var output: String?
    public var token = false
    public var json = false
    public var globals = false

    public init() {}
}

public enum SnapshotCommand {
    /// `envdoctor snapshot` — capture this machine's live runtime.
    public static func run(_ args: SnapshotArgs) throws -> Int {
        let root = resolveRoot(args.root)
        let snapshot = Capture.captureSnapshot(globals: args.globals)

        if let output = args.output {
            let dest = resolvePath(root, output)
            try writeFile(at: dest, content: prettySnapshot(snapshot) + "\n")
            FileHandle.standardError.write("✓ Snapshot written to \(output)\n")
        }

        if args.json {
            print(prettySnapshot(snapshot))
            return 0
        }

        if args.token {
            print(Token.encodeToken(snapshot))
            return 0
        }

        // Human summary.
        let title = "RUNTIME SNAPSHOT"
        print(title)
        print(rule(title.count * 2) + "\n")
        print("  OS  \(snapshot.osPlatform)/\(snapshot.osArch) \(snapshot.osRelease)\n")

        print("  Tools")
        if snapshot.tools.isEmpty {
            print("  none detected")
        } else {
            for t in snapshot.tools {
                print("  ✓ \(padEnd(t.tool, 8)) \(t.version)  \(t.resolvedFrom)")
            }
        }

        print("\n  PATH (\(snapshot.path.count) entries)")
        for (i, p) in snapshot.path.prefix(12).enumerated() {
            print("  \(padStart(String(i + 1), 2))  \(p)")
        }
        if snapshot.path.count > 12 {
            print("  … \(snapshot.path.count - 12) more")
        }

        if !snapshot.globals.isEmpty {
            print("\n  Globals")
            for (eco, packages) in snapshot.globals {
                print("  \(eco): \(packages.count) packages")
            }
        } else if !args.globals {
            print("\n  Globals omitted — pass --globals to include the package inventory.")
        }

        print("\n  Share with:  envdoctor snapshot --token   ·   compare with:  envdoctor snapshot-diff <a> <b>")

        return 0
    }

    /// The full snapshot as pretty JSON (`JSON.stringify(snapshot, null, 2)`
    /// key order).
    public static func prettySnapshot(_ snapshot: RuntimeSnapshot) -> String {
        Json.pretty(snapshotJsonValue(snapshot))
    }

    static func snapshotJsonValue(_ snapshot: RuntimeSnapshot) -> JSONValue {
        var globals = JSONObject()
        for (eco, packages) in snapshot.globals {
            globals.set(eco, .array(packages.map { p in
                var pkg = JSONObject()
                pkg.set("name", .string(p.name))
                pkg.set("version", .string(p.version))
                return .object(pkg)
            }))
        }
        var osObj = JSONObject()
        osObj.set("platform", .string(snapshot.osPlatform))
        osObj.set("arch", .string(snapshot.osArch))
        osObj.set("release", .string(snapshot.osRelease))

        var root = JSONObject()
        root.set("schema", .int(snapshot.schema))
        root.set("capturedAt", .string(snapshot.capturedAt))
        root.set("os", .object(osObj))
        root.set("tools", .array(snapshot.tools.map { t in
            var tool = JSONObject()
            tool.set("tool", .string(t.tool))
            tool.set("version", .string(t.version))
            tool.set("resolvedFrom", .string(t.resolvedFrom))
            return .object(tool)
        }))
        root.set("path", .array(snapshot.path.map { .string($0) }))
        root.set("globals", .object(globals))
        root.set("envFlagNames", .array(snapshot.envFlagNames.map { .string($0) }))
        return .object(root)
    }
}
